test/poseidon.cpp

```diff
@@ -92,102 +97,221 @@ void hash_test_data(const TestSet &test_set) {
+                                                0x1A842A688E600F012637FE181292F70C4347B5AE0D9EA9CE7CF18592C345CF73_cppui256};
+    field_type::value_type expected_result = 0x3F4B0EABB64E025F920457AF8D090A9F6472CAE11F3D62A749AF544A44941B9B_cppui256;
+    sponge_construction_type pallas_sponge;
+    
+    pallas_sponge.absorb(input);
+    field_type::value_type res = pallas_sponge.squeeze();
+
+    BOOST_CHECK(res == expected_result);
+}
+
+BOOST_AUTO_TEST_CASE(poseidon_kimchi_test_4) {
+    using field_type = fields::pallas_base_field;
+    using poseidon_policy = hashes::detail::base_poseidon_policy<field_type, 2, 1, 7, 55, 0, true>;
+    using sponge_construction_type = hashes::detail::poseidon_sponge_construction<poseidon_policy>;
+
+    std::vector<field_type::value_type> input = {0x2059462D60621F70620EA697FA1382EC5553A3DADB3CF9072201E09871B8284C_cppui256,
+                                                0x2747337D1C4F9894747074C771E8EC7F570640E5D0CAF30FDDC446C00FA48707_cppui256,
+                                                0x2DD5047C3EEEF37930E8FA4AD9691B27CF86D3ED39D4DEC4FC6D4E8EE4FF0415_cppui256,
+                                                0x12C387C69BDD436F65AB607A4ED7C62714872EDBF800518B58E76F5106650B29_cppui256};
+    field_type::value_type expected_result = 0x165A8CECF6660C6E0054CB9B4DBA9D68047166D7F3CED2F8DC86ED2EBFD3EC47_cppui256;
+    sponge_construction_type pallas_sponge;
+    
+    pallas_sponge.absorb(input);
+    field_type::value_type res = pallas_sponge.squeeze();
+
+    BOOST_CHECK(res == expected_result);
+}
+
+BOOST_AUTO_TEST_CASE(poseidon_kimchi_test_5) {
+    using field_type = fields::pallas_base_field;
+    using poseidon_policy = hashes::detail::base_poseidon_policy<field_type, 2, 1, 7, 55, 0, true>;
+    using sponge_construction_type = hashes::detail::poseidon_sponge_construction<poseidon_policy>;
+
+    std::vector<field_type::value_type> input = {0x3CF70C3A89749A45DB5236B8DE167A37762526C45270138A9FCDF2352B1899DA_cppui256,
+                                                0x1BDF55BC84C1A0E0F7F6834949FCF90279B9D21C17DBC9928202C49039570598_cppui256,
+                                                0x09441E95A82199EFC390152C5039C0D0566A90B7F6D1AA5813B2DAB90110FF90_cppui256,
+                                                0x375B4A9785503C24531723DB1F31B50B79C3D1EC9F95DB7645A3EDA03862B588_cppui256,
+                                                0x12688FE351ED01F3BB2EB6B0FA2A70FB232654F32B08990DC3A411E527776A89_cppui256};
+    field_type::value_type expected_result = 0x0CA2C3342C2959D7CD94B5C9D4DC55900F5F60B345F714827C8B907752D5A209_cppui256;
+    sponge_construction_type pallas_sponge;
+    
+    pallas_sponge.absorb(input);
+    field_type::value_type res = pallas_sponge.squeeze();
+
+    BOOST_CHECK(res == expected_result);
+}
+
+/*BOOST_DATA_TEST_CASE(poseidon_strengthen_1, string_data("internal", "strengthen", "1"), data_set) {
```

Comment:
Because of how we set the kimchi constants, we cannot run the usual Poseidon.

Reply:
I have refactored most of the code, please review again.